daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,242 @@
+.. Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+V1 Contract ID allocation scheme
+================================
+
+Goals
+^^^^^
+
+* Allows ordering contract IDs and make DAML semantics depend on this
+  order, e.g., for comparison builtin and maps using IDs as keys.
+* Eliminate all contract ID translations for central commiter ledger
+* Allows ledgers to store arbitrary information about the contract and the creating transaction in the contract ID if necessary.
+
+Requirements
+^^^^^^^^^^^^
+
+**Executability**: DAML engine can determine ordering of the contract
+IDs before it hands out the transaction to the write service.
+
+**Validation**: The allocation scheme commutes with transaction
+projection. That is, the contract IDs for contracts created in the
+projection of a transaction to a set of parties can be computed
+solely from the projection and input seeds.
+
+**Unlinkability**: It is computationally infeasible to link the contract
+contents to the contract ID unless the create node is witnessed or the input seeds are known. The
+contract contents include the contract instance, the template ID, the
+stakeholders / signatories / maintainers, and the contract key.
+
+**Freshness**: It is computationally infeasible to find contracts C1 and
+C2 with the same contract ID such that one of the following holds:
+
+* The contracts have different stakeholders.
+* The contracts are created by different transactions.
+* The contracts are created by different subactions of the same
+  transaction forest where the seeds are pairwise different.
+
+**Distinctness**: The ledger can enforce that the seeds are pairwise
+different.
+
+Contract Identifiers
+^^^^^^^^^^^^^^^^^^^^
+
+A *contract identifier* (or contract ID for short) is a sequence of
+bytes defined as follows ::
+
+  ContractID := verisionPrefix ∥ discriminator ∥ suffix  
```

Comment:
Should it be `versionPrefix`? If so, the same type happened further below again.